mountpoint-s3-fs/src/superblock/handles_map.rs

```diff
@@ -0,0 +1,160 @@
+use crate::metablock::InodeNo;
+use crate::sync::Mutex;
+use std::collections::HashMap;
+use std::collections::hash_map::Entry;
+
+use super::InodeLockedForWriting;
+
+/// Tracks currently open readers and writers (file handles) for Inodes.
+/// Ensures that the Inodes are locked for writing while performing any operation.
+#[derive(Debug, Default)]
+pub struct InodeHandleMap {
+    handles: Mutex<HashMap<InodeNo, InodeHandles>>,
+}
+
+#[derive(Debug, Default)]
+struct InodeHandles {
+    readers: HashMap<u64, HandleState>,
+    writer: Option<(u64, HandleState)>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum HandleState {
+    Active,
+    Inactive,
+}
+
+impl InodeHandles {
+    fn is_empty(&self) -> bool {
+        self.readers.is_empty() && self.writer.is_none()
+    }
+}
+
+impl InodeHandleMap {
+    /// Add a new active reader. Will fail if there is an active writer.
+    pub fn try_add_reader(&self, locked_inode: &InodeLockedForWriting<'_>, fh: u64) -> bool {
+        let mut handles = self.handles.lock().unwrap();
+        let entry = handles.entry(locked_inode.ino).or_default();
+        if entry
+            .writer
+            .as_ref()
+            .is_some_and(|(_, state)| *state == HandleState::Active)
+        {
+            return false;
+        }
+        entry.writer = None;
+        entry.readers.insert(fh, HandleState::Active);
+        true
+    }
+
+    /// Set an existing reader to inactive.
+    pub fn try_deactivate_reader(&self, locked_inode: &InodeLockedForWriting<'_>, fh: u64) {
+        let mut handles = self.handles.lock().unwrap();
+        if let Some(entry) = handles.get_mut(&locked_inode.ino)
+            && let Some(reader) = entry.readers.get_mut(&fh)
+        {
+            *reader = HandleState::Inactive;
+        }
+    }
+
+    /// Set an existing reader to active.
+    ///
+    /// Return `false` if the reader did not exist.
+    pub fn try_activate_reader(&self, locked_inode: &InodeLockedForWriting<'_>, fh: u64) -> bool {
+        let mut handles = self.handles.lock().unwrap();
+        if let Some(entry) = handles.get_mut(&locked_inode.ino)
+            && let Some(reader) = entry.readers.get_mut(&fh)
```

Comment:
if `reader` is not found it's unexpected, right? should we report it as an error?

I am thinking of a case when `fs::read` happens and `flushed == true` there

Reply:
Do we also need to check that there is no writer: `&& entry.writer.is_none()`?

Otherwise, I think in this case we will allow the read:
    
```
fd1 = open("a.txt", "r")
fd2 = dup(fd1)
close(fd1) // flush, but no release yet (deactivated, but still in the `entry.readers` map)
fd3 = open("a.txt", "w") // override (all readers are inactive)
read(fd2) // `try_reactivate_handle` succeeds and the read is allowed (while there is an unflushed writer!)
```

Or am I missing something and something in our code prohibits such scenario? Note that some data may be written to `fd3` before the `read(fd2)`.